### DISCUSSION TITLE
Template rendering

### DISCUSSION THREAD

**Discussion by JoshPolaris** [marblejs/marble#315] [Q&A] [answered]
Recently learned about Marblejs and would really love to use it, but my work is not usually about building REST APIs. I know Marblejs is geared towards APIs, but was wondering if there's a solution on rendering templates.

**Comment by JozefFlakus** [comment, marked as answer]
Hi @JoshPolaris,
Could you describe more the use case in which you would like to use server side rendering?
Marble.js doesn't have any built-in mechanism (module/package) for rendering templates but it doesn't mean that it is impossible. It is just a typical HTTP flow. Generally speaking it is a matter of integrating a specific template rendering engine, like eg. [mustache.js](https://github.com/janl/mustache.js/) or something else, even React or other library. Nothing too complex, really. :)


```typescript
export const foo$ = r.pipe(
  r.matchPath('/foo'),
  r.matchType('GET'),
  r.useEffect((req$, ctx) => {
    // lets assume that there is an already registered renderer instance in the context,
    // eg. the previously mentioned mustache.js
    const htmlRenderer = useContext(SomeHtmlRendererToken)(ctx.ask);

    return req$.pipe(
      mergeMap(flow(
        extractParameters,
        htmlRenderer.render)),
      map(body => ({
        status: HttpStatus.OK,
        headers: { 'Content-Type': 'text/html' },
        body,
      });
    );
  }),
);
```

**update:**
Additionally to this you have to expose an endpoint for serving static files, so eg. if you reference any file like, css, js, images, etc. you have to register an effect which will be able to serve those files.

```typescript
const getFile$ = r.pipe(
  r.matchPath('/static/:dir*'),
  r.matchType('GET'),
  r.useEffect(req$ => req$.pipe(
    validateRequest,
    map(req => req.params.dir),
    mergeMap(readFile),
    map(body => ({ body })),
  )),
);
```
